File: solutions/871-keys-and-rooms/keys-and-rooms.cpp
class Solution {
public:
    bool canVisitAllRooms(std::vector<std::vector<int>>& rooms) {
        int n = rooms.size();
        std::vector<bool> vis(n, false);
        std::stack<int> stack;
        stack.push(0);
        vis[0] = true;
        int count = 1;

        while (!stack.empty()) {
            int currentRoom = stack.top();
            stack.pop();
            for (int key : rooms[currentRoom]) {
                if (!vis[key]) {
                    stack.push(key);
                    vis[key] = true;
                    count++;
                }
            }
        }
        return count == n;
    }
};

File: solutions/871-keys-and-rooms/keys-and-rooms.py
class Solution:
    def canVisitAllRooms(self, rooms: List[List[int]]) -> bool:
        n = len(rooms)
        vis = [False] * n
        stack = [0]
        vis[0] = True
        count = 1

        while stack:
            current_room = stack.pop()
            for key in rooms[current_room]:
                if not vis[key]:
                    stack.append(key)
                    vis[key] = True
                    count += 1
        
        return count == n